osu.Game/Database/DatabaseContextFactory.cs

```diff
@@ -1,27 +1,98 @@
 ﻿// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
 // Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
 
+using System.Threading;
 using osu.Framework.Platform;
 
 namespace osu.Game.Database
 {
-    public class DatabaseContextFactory
+    public class DatabaseContextFactory : IDatabaseContextFactory
     {
         private readonly GameHost host;
 
         private const string database_name = @"client";
 
+        private ThreadLocal<OsuDbContext> threadContexts;
+
+        private readonly object writeLock = new object();
+
+        private OsuDbContext writeContext;
+
+        private bool currentWriteDidWrite;
+        private volatile int currentWriteUsages;
+
         public DatabaseContextFactory(GameHost host)
         {
             this.host = host;
+            recycleThreadContexts();
+        }
+
+        /// <summary>
+        /// Get a context for read-only usage.
+        /// </summary>
+        public OsuDbContext Get() => threadContexts.Value;
+
+        /// <summary>
+        /// Request a context for write usage. Can be consumed in a nested fashion (and will return the same underlying context).
+        /// This method may block if a write is already active on a different thread.
+        /// </summary>
+        /// <returns>A usage containing a usable context.</returns>
+        public DatabaseWriteUsage GetForWrite()
+        {
+            Monitor.Enter(writeLock);
+
+            Interlocked.Increment(ref currentWriteUsages);
+
+            return new DatabaseWriteUsage(writeContext ?? (writeContext = threadContexts.Value), usageCompleted);
+        }
+
+        private void usageCompleted(DatabaseWriteUsage usage)
+        {
+            int usages = Interlocked.Decrement(ref currentWriteUsages);
+
+            try
+            {
+                currentWriteDidWrite |= usage.PerformedWrite;
+
+                if (usages > 0) return;
+
+
+                if (currentWriteDidWrite)
+                {
+                    writeContext.Dispose();

```

Comment:
Why is this explicit `Dispose` needed?

Reply:
because we should be disposing these where we can. after an import the context can be very large.

Reply:
Hmmm... Alright. I would've really liked to rely on the GC but if the contexts are so large that they would cause significant stutters if executed elsewhere, I guess this is the better location for it.